"e2e.test --help" panics due to bad flag-parsing code

#### What happened:
I ran `e2e.test --help` and it panicked:

```
> ./_output/bin/e2e.test --help
Usage of ./_output/bin/e2e.test:
  -add_dir_header
    	If true, adds the file directory to the header of the log messages
...
  -gce-zone string
    	GCE zone being used, if applicable
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1802967]

goroutine 1 [running]:
k8s.io/kubernetes/vendor/k8s.io/component-base/cli/flag.(*StringSlice).String(0xc00399d720, 0x4547c40, 0xc00399d720)
	/home/danw/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/component-base/cli/flag/string_slice_flag.go:41 +0x27
...
```

#### What you expected to happen:
no panic

#### How to reproduce it (as minimally and precisely as possible):
```
make WHAT=test/e2e/e2e.test
./_output/bin/e2e.test --help
```

#### Anything else we need to know?:
The panic was introduced by #98787. I submitted #99704 as an attempt to fix this, but review led to the realization that there were further bugs and I don't know what the right fix is

cc @DangerOnTheRanger

--- a/staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go
+++ b/staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go
@@ -38,12 +38,16 @@ var _ goflag.Value = &StringSlice{}
 var _ pflag.Value = &StringSlice{}
 
 func (s *StringSlice) String() string {
+	if s == nil || s.value == nil {
+		return ""
+	}
 	return strings.Join(*s.value, " ")
 }
 
 func (s *StringSlice) Set(val string) error {
 	if s.value == nil || !s.changed {
-		*s.value = make([]string, 0)
+		v := make([]string, 0)
+		s.value = &v
 	}
 	*s.value = append(*s.value, val)
 	s.changed = true

--- a/staging/src/k8s.io/component-base/cli/flag/string_slice_flag_test.go
+++ b/staging/src/k8s.io/component-base/cli/flag/string_slice_flag_test.go
@@ -83,8 +83,8 @@ func TestStringSlice(t *testing.T) {
 		} else if err != nil {
 			t.Errorf("%d: expected nil error, got %v", i, err)
 		}
-		if !reflect.DeepEqual(s, test.expected) {
-			t.Errorf("%d: expected %+v, got %+v", i, test.expected, s)
+		if !reflect.DeepEqual(*v.value, test.expected) {
+			t.Errorf("%d: expected %+v, got %+v", i, test.expected, *v.value)
 		}
 		if v.changed != test.changed {
 			t.Errorf("%d: expected %t got %t", i, test.changed, v.changed)

